File: Data Structures & Algorithms/search-2d-matrix/submission-4.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        def searchRow(target: int) -> int:
            up, down = 0, len(matrix) - 1
            while up <= down:
                mid = (up + down) // 2
                if matrix[mid][0] > target:
                    down = mid - 1
                elif matrix[mid][-1] < target:
                    up = mid + 1
                elif matrix[mid][0] <= target and matrix[mid][-1] >= target:
                    return mid
            return -1

        row = searchRow(target)
        if row == -1: 
            return False
        
        nums = matrix[row]
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = (l + r) // 2
            if nums[mid] == target:
                return True
            elif nums[mid] > target:
                r = mid - 1
            else:
                l = mid + 1
        return False
                    

File: Data Structures & Algorithms/minimum-size-subarray-sum/submission-4.cpp
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums) {
        int r = 0, l = 0;
        int res = nums.size() + 1, len = 0;
        int sum = 0;
        while (r < nums.size()) {
            sum += nums[r];
            while (sum >= target) {
                res = min(res, r - l + 1);
                sum -= nums[l++];
            } 
            r++;
        }
        return (res == nums.size() + 1) ? 0 : res;
    }
};